internal/proxy/client.go

```diff
@@ -1,42 +1,218 @@
 	switch addr {
 	case "kvstore":
-		return abciclient.NewLocalCreator(kvstore.NewApplication()), noopCloser{}
+		return abciclient.NewLocalClient(logger, kvstore.NewApplication()), noopCloser{}, nil
 	case "persistent_kvstore":
 		app := kvstore.NewPersistentKVStoreApplication(logger, dbDir)
-		return abciclient.NewLocalCreator(app), app
+		return abciclient.NewLocalClient(logger, app), app, nil
 	case "e2e":
 		app, err := e2e.NewApplication(e2e.DefaultConfig(dbDir))
 		if err != nil {
-			panic(err)
+			return nil, noopCloser{}, err
 		}
-		return abciclient.NewLocalCreator(app), noopCloser{}
+		return abciclient.NewLocalClient(logger, app), noopCloser{}, nil
 	case "noop":
-		return abciclient.NewLocalCreator(types.NewBaseApplication()), noopCloser{}
+		return abciclient.NewLocalClient(logger, types.NewBaseApplication()), noopCloser{}, nil
 	default:
 		mustConnect := false // loop retrying
-		return abciclient.NewRemoteCreator(logger, addr, transport, mustConnect), noopCloser{}
+		client, err := abciclient.NewClient(logger, addr, transport, mustConnect)
+		if err != nil {
+			return nil, noopCloser{}, err
+		}
+
+		return client, noopCloser{}, nil
 	}
 }
 
 type noopCloser struct{}
 
 func (noopCloser) Close() error { return nil }
+
+////////////////////////////////////////////////////////////////////////
+
+// proxyClient implements provides the application connection.
+type proxyClient struct {
+	service.BaseService
+	logger log.Logger
+
+	client  abciclient.Client
+	metrics *Metrics
+}
+
+// New creates a proxy application interface.
+func New(client abciclient.Client, logger log.Logger, metrics *Metrics) abciclient.Client {
+	conn := &proxyClient{
+		logger:  logger,
+		metrics: metrics,
+		client:  client,
+	}
+	conn.BaseService = *service.NewBaseService(logger, "proxyClient", conn)
+	return conn
+}
+
+func (app *proxyClient) OnStop()      { tryCallStop(app.client) }
+func (app *proxyClient) Error() error { return app.client.Error() }
+
+func tryCallStop(client abciclient.Client) {
+	switch c := client.(type) {
+	case nil:
+		return
+	case interface{ Stop() }:
+		c.Stop()
+	}
+}
+
+func (app *proxyClient) OnStart(ctx context.Context) error {
+	var err error
+	defer func() {
+		if err != nil {
+			tryCallStop(app.client)
+		}
+	}()
+
+	// Kill Tendermint if the ABCI application crashes.
+	go func() {
+		if !app.client.IsRunning() {
+			return
+		}
+		app.client.Wait()
+		if ctx.Err() != nil {
+			return
+		}
+
+		if err := app.client.Error(); err != nil {
+			app.logger.Error("client connection terminated. Did the application crash? Please restart tendermint",
+				"err", err)
+
+			if killErr := kill(); killErr != nil {
+				app.logger.Error("Failed to kill this process - please do so manually",
+					"err", killErr)
+			}
+		}
+
+	}()
+
+	return app.client.Start(ctx)
+}
+
+func kill() error {
+	p, err := os.FindProcess(os.Getpid())
+	if err != nil {
+		return err
+	}
+
+	return p.Signal(syscall.SIGTERM)
+}
```

Comment:
This seems a little rough. It seems like we really want to communicate up to whatever system is managing the services that this particular service died and that the reason it died puts the rest of the system in a critical condition.

I think a reasonably straightforward way to accomplish this would be to add an error type to the `service` package that communicates a critical issue:
```
type FatalError struct{
    Reason string
}
````
And add an error return to the `Wait` service method. 

The current implementation basically means that, an logic that wants to persistent in the event of an ABCI application crash can't and instead has to die. I can imagine this making tests a bit more confusing as well, since tests that crash the ABCI application or don't shutdown in the correct order will, I believe, kill the go test process.

Reply:
I think this is probably more than a little rough, and I think the right thing to do is to shut down (or panic) rather than signaling ourselves. This definitely feels like the kind of exceptional situation where, if there's an error condition in our client that we should halt as quickly as possible. Adding an error return to wait would be kind of a big change, with limited utility: we don't currently track any errors during shut down (waiting or just calling stop,) and there are a *lot* of services that we'd want to change if that were the case.

The code has always been like this, so the fact that we don't hit this *all the time* in normal operation (which looking at the code is what I expect) makes me think that it's not a huge risk as is. This is a thing that Is bad (though I think it's less bad with this patch than before), but despite being bad it continues to not feel like a huge concern for me. 

Increasingly, I think we'll get rid of the proxy package entirely, and I suspect this will fall away then.

Reply:
Yeah. I thought about making more or less the same comment @williambanfield did, but I realized that we don't really have an orchestrator in most deployments apart from `init`, and this may not be ideal, but as a workaround it's not bad. 

There's a whole philosophical discussion around graceful shutdown, and I'm firmly in the "don't bother the cure is worse than the disease" camp, but the truth is there's no universal answer.

Reply:
> Yeah. I thought about making more or less the same comment @williambanfield did, but I realized that we don't really have an orchestrator in most deployments apart from `init`, and this may not be ideal, but as a workaround it's not bad.
> 
> There's a whole philosophical discussion around graceful shutdown, and I'm firmly in the "don't bother the cure is worse than the disease" camp, but the truth is there's no universal answer.

Hah, yeah, I think as a person I'm definitely in the camp that we should work to be safer than average here (because I can envision leaving the DB in a mixed state, and that seems bad,) and I think we'll have to reopen a lot of these questions if/when we start doing multi-tenancy (multiple tendermints in one process,) but i think this is all a step in the right direction.